Context file editor (frontend bit)

matheuscteo opened the pull request:
### Details

#### 1. **Editor Slice**

- Editor state:  isOpen, lastModified and unseenChanges
- Reducers: openEditor, resetEditor, upadateLastModified, clearUnseenChanges.

#### 2. **Editor API**

- RTK Query endpoints for fetching file content and last-modified timestamp.

#### 3. **Editor Component**

- Monaco editor

#### 4. **EditorTab Component**

- Displays the editor inside a panel.
- Shows the last updated timestamp at the bottom.
- Uses ReduxTK to get the latest modification time.

#### 5. **Dashboard Integration**

- Added a button to open the context file (Editor tab) in the main tabs.
- Button color changes if there are unseen changes.
- Polls for file changes and updates the Redux state.
- Ensures the Editor tab is closable and state is managed correctly.


---

### Screenshots

![Screenshot 2025-05-05 at 09 27 12](https://github.com/user-attachments/assets/4df5d378-2dbb-4419-9617-68d58d1ab3a5)
![Screenshot 2025-05-05 at 09 27 29](https://github.com/user-attachments/assets/c9ad626b-b2bd-44bf-8582-f2521875c5c5)
![Screenshot 2025-05-05 at 09 30 53](https://github.com/user-attachments/assets/dad61d33-09b7-4868-be2a-c4535d4a38da)


---

### How to Test

1. Open the dashboard.
2. Click the "Context file" button to open the Editor tab. Check the content.
3. Close the Editor tab or select a different tab so the Editor tab isn't in view.
4. Make changes to the context file externally and observe the "unseen changes" indicator.
5. Check that the last updated timestamp updates as expected so as the content.

CammilleCC commented:
Thanks for the contribution, it looks really pretty 🤩 

First impressions on the UI:
- Maybe we should pin the context file tab, similar to the PyQt GUI. It might be good to have a familiar look and feel across interfaces.
- I am unsure how beneficial having the timestamp of the current context file on the overview. Maybe having an LED beside the context file tabs (e.g., like the unsaved changes in VS Code) to denote the status?

Could you also please run ESLint and Prettier on the frontend codebase?

matheuscteo commented:
> Thanks for the contribution, it looks really pretty 🤩
> 
> First impressions on the UI:
> 
> * Maybe we should pin the context file tab, similar to the PyQt GUI. It might be good to have a familiar look and feel across interfaces.
> * I am unsure how beneficial having the timestamp of the current context file on the overview. Maybe having an LED beside the context file tabs (e.g., like the unsaved changes in VS Code) to denote the status?
> 
> Could you also please run ESLint and Prettier on the frontend codebase?



Thanks for the feedback.

I have no strong opinion on Pinning the context file tab. Also if we go for it I can remove the eye-shaped-button too and inform the changes via the dot in the tab, sure. That's an option but I'd like to ponder a bit more on that:

- 1 - Too many tabs that might have no meaning to the user. Imagine the user opens a plot tab and in the future a view tab and so on. As I learned from some instrument scientists, many users aren't interested in the context file per said (hence the whole discussion about limiting the editing to just a set of users per proposal). If not important for all users -> possible to close.

- 2 - New comers to DAMNIT might not really get the LED strategy since they might have never seen the PyQt version before. That's why I went for the full timestamp displaying.
  - 2.1 -  Now that I see it again, having a full timestamp display is surely excessive. Perhaps one can add a tool tip to the eye-shaped-button with the timestamp.
  - 2.2 - If we go for the LED and fixed tab, I'd not even add in the tool-tip for we assume the user knows what's about.

CammilleCC commented:
>     * 1 - Too many tabs that might have no meaning to the user. Imagine the user opens a plot tab and in the future a view tab and so on. As I learned from some instrument scientists, many users aren't interested in the context file per said (hence the whole discussion about limiting the editing to just a set of users per proposal). If not important for all users -> possible to close.

We don't have much tabs yet to worry about this, and we should consider settling users gradually with the new interface by exposing features in a familiar way.

The Plots tab actually was not foreseen as its main entrypoint, it is planned for them to be opened in a new window for side-by-side comparison, same with the context file editor. 

We could revisit optional closing in the future, especially if we would support multiple table views.

>     * 2 - New comers to DAMNIT might not really get the LED strategy since they might have never seen the PyQt version before. That's why I went for the full timestamp displaying.

Actually I was thinking something more like this:
![image](https://github.com/user-attachments/assets/94430f03-02b7-4b73-8a59-281ce8e44671)

I believe having a minimal status indicator for a noncritical feature, as you have mentioned, should be reduced to minimize focus.
The timestamp can be displayed on the editor tab instead, where it probably where that information is more relevant.

matheuscteo commented:
![Screenshot 2025-05-07 at 08 50 21](https://github.com/user-attachments/assets/9c5d17a5-6449-43e4-be4e-3715c9365b90)

I added the LED and removed the eye-shaped button.

takluyver commented:
The LED indicator you've made looks neat, but I think from the code it will be shown for *unseen* changes, i.e. if someone else has edited the file and it has been automatically updated in the editor? This will be pretty common for a lot of experiments, and we don't expect everyone with DAMNIT open to go to the editor tab to look at the latest changes (if they can even find them). So I think we probably don't want such a prominent, persistent indicator for that.

The indicator in the Qt GUI - the tab title turns bold and red - is for *unsaved* changes, i.e. you have changed the context file but not yet saved. I think it's common for text editors to indicate that state, so it's hopefully familiar even to people who haven't used our Qt GUI.

Also, looking at the code, I think the LED is blinking (`animation: "blink 1s infinite alternate",`)? Anything animated is pretty attention grabbing, so let's keep it for things where it's quite important to get the user's attention, or maybe to suggest something is in progress (as in spinners).

matheuscteo commented:
> The LED indicator you've made looks neat, but I think from the code it will be shown for _unseen_ changes, i.e. if someone else has edited the file and it has been automatically updated in the editor? This will be pretty common for a lot of experiments, and we don't expect everyone with DAMNIT open to go to the editor tab to look at the latest changes (if they can even find them). So I think we probably don't want such a prominent, persistent indicator for that.
> 
> The indicator in the Qt GUI - the tab title turns bold and red - is for _unsaved_ changes, i.e. you have changed the context file but not yet saved. I think it's common for text editors to indicate that state, so it's hopefully familiar even to people who haven't used our Qt GUI.
> 
> Also, looking at the code, I think the LED is blinking (`animation: "blink 1s infinite alternate",`)? Anything animated is pretty attention grabbing, so let's keep it for things where it's quite important to get the user's attention, or maybe to suggest something is in progress (as in spinners).

I've removed the LED and added a fontWeight and a color (orange so it's not that distracting as red, I guess?)

![image](https://github.com/user-attachments/assets/900f473a-cf8b-4055-98bd-6f895ecd3b8a)

CammilleCC commented on frontend/src/features/editor/editor-tab.tsx:
```diff
@@ -0,0 +1,102 @@
+          <Editor fileContent={data?.fileContent} error={error} />
+        </Paper>
+      </Box>
+      <Box
```
I would suggest making a component for this status indicator.

CammilleCC reviewed (commented):
So far so good! Please run `eslint` for some TS and linting improvements.

CammilleCC commented:
Merging this branch on behalf of @matheuscteo :blush: good addition~
Just added some small fixes to make the TypeScript build pass, would be good to revisit the open suggestions in the future (though minor and nitpicky).
